psyneulink/core/llvm/codegen.py

```diff
@@ -260,6 +350,44 @@ def visit_Name(self, node):
 
     def visit_Attribute(self, node):
         val = self.visit(node.value)
+
+        # special case numpy attributes
+        if node.attr == "shape":
+            # recursively extract shape from value
+            dimensions = []
+            val = val.type.pointee
+            while hasattr(val, "count"):
+                dimensions.append(val.count)
+                val = val.element
+
+            # dimensions is a tuple, so it should be a llvm struct
+            dimensions_struct = self.builder.alloca(ir.LiteralStructType([self.ctx.float_ty] * len(dimensions)))
+            for idx, dim in enumerate(dimensions):
+                struct_ptr = self.builder.gep(dimensions_struct, [self.ctx.int32_ty(0), self.ctx.int32_ty(idx)])
+                self.builder.store(self.ctx.float_ty(dim), struct_ptr)
+            return dimensions_struct
+        elif node.attr == "astype":
+            def astype(ty):
+                def _convert(ctx, builder, x):
+                    if helpers.is_pointer(x):
+                        x = builder.load(x)
+                    if helpers.is_integer(x) and ty is ctx.float_ty:
+                        if helpers.is_boolean(x):
+                            return builder.uitofp(x, ty)
+                        return builder.sitofp(x, ty)
+                    elif helpers.is_floating_point(x) and ty is self.ctx.float_ty:
```

Comment:
this case (float -> int) actually turns out to be difficult to test for, since return types from compiled udfs are always assumed to be doubles.
in addition, the way I detect size of the integers might result in different int sizes than the rest of the compiled code

it may be worth it to not allow float -> int conversions altogether if this is the case

Reply:
imo, it's good for now. it can be revisited later.